proc: bugfix: Truncate stacktrace when FDE of a frame can not be found

aarzilli opened the pull request:
Instead of returning an error when FDE of a frame can not be found,
just truncate the stack trace.

Fixes #462

I also did a bit of refactoring of the stacktrace code:

- made GoroutineStacktrace a method of struct G
- made stacktrace a method of StackIterator
- renamed StackIterator to stackIterator
- factored out logic to obtain a stackIterator from a goroutine that's
used by both (*G).Stacktrace and by (*G).UserCurrent

derekparker commented:
@aarzilli what's the underlying cause of this issue? Why do we encounter PCs we cannot reconcile with an FDE? I haven't dug into this, but is it a bug in our FDE lookup, or in the information generated?

aarzilli commented:
`runtime.mcall` (which happens to be at the bottom of the stack in the related issue) is apparently one of those functions that never returns. In general I think it's better to show a truncated stack trace rather than a error message when we can't decode a stack frame.

derekparker commented:
@aarzilli makes sense, but in that case we should print a warning that the stack trace has been truncated. Additionally, I think that `runtime.mcall` should be considered a "top of stack" function, which should signify that we can stop collecting frames.

aarzilli commented:
Ok, I've changed `proc.(*stackIterator).Next` so that it will return an extra frame with `File` "?" and `Line` -1 for all undecodable frames except when the undecodable frame is the top of the stack (in that case we lose no informations in returning the error directly).
A truncated stack trace will look like this:

```
(dlv) stack
0  0x000000000045d6a9 in runtime.epollwait
   at /usr/local/go/src/runtime/sys_linux_amd64.s:440
1  0x0000000000428acc in runtime.netpoll
   at /usr/local/go/src/runtime/netpoll_epoll.go:67
2  0x00000000004323e3 in runtime.findrunnable
   at /usr/local/go/src/runtime/proc.go:1955
3  0x0000000000432b7a in runtime.schedule
   at /usr/local/go/src/runtime/proc.go:2072
4  0x0000000000432e16 in runtime.park_m
   at /usr/local/go/src/runtime/proc.go:2137
5  0x0000000000459b09 in runtime.mcall
   at /usr/local/go/src/runtime/asm_amd64.s:233
6  0x000000000091e800 in (nil)
   at ?:-1
```

Additionally I have added runtime.mcall to the "top of stack" functoins.

aarzilli commented:
Found a bug while investigating issue #468 (stacktrace returned 0 stack frames and no errors because PCToLine didn't find anything), I've added a commit to address it.

aarzilli commented:
I've disabled the test on windows since [there are no stack frames to print](https://ci.appveyor.com/project/derekparker/delve/build/254).

derekparker commented:
Merged.
